### PR TITLE
Add tap pmd

### PR DESCRIPTION
**Pull request by KeithWiles**
tap:add new tun/tap PMD support

Create a new PMD for tun/tap type devices. The PMD needs to have some
performance updates nd more features at some point, but for now it is
reasonable.

Signed-off-by: Keith Wiles <keith.wiles@intel.com>

### CODE REVIEW

File: lib/core/pmds/net/tap/pmd_tap.c [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+        for (int i = 0; i < nb_pkts; i++) {
+            pktmbuf_t *m = bufs[i];
+            ssize_t len;
+
+            pi.flags = 0;
+            pi.proto = 0;
+            if (tap_type == IFF_TUN) {
+                char proto = *pktmbuf_mtod(m, char *);
+
+                if (proto == 0x40)
```
Comment (jeffreybshaw) [inline lib/core/pmds/net/tap/pmd_tap.c:146]: Is the IP protocol version supposed to be set in teh packet data? Is that what you're doing here? Shouldn't you mask off the version bits?

File: lib/core/pmds/net/tap/pmd_tap.c [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+    dev_info->if_index    = if_nametoindex(lport->if_name);
+
+    return 0;
+}
+
+static int
+pmd_stats_get(struct cne_pktdev *dev, lport_stats_t *stats)
+{
+    struct pmd_lport *lport = dev->data->dev_private;
+    struct tap_rx_q *rxq    = lport->rxq;
```
[the remark this answers is not in the document]
Comment (KeithWiles) [reply lib/core/pmds/net/tap/pmd_tap.c:212]: This code was taken from af_packet, so we need to fix that PMD too.

[the remark this answers is not in the document]
Comment (KeithWiles) [reply lib/core/pmds/net/tap/pmd_tap.c:212]: Need to update af_packet PMD too.

File: lib/core/pmds/net/tap/pmd_tap.c [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+    return pmd_tuntap_tx(queue, bufs, nb_pkts, IFF_TUN);
+}
+
+static int
+pmd_tap_dev_info(struct cne_pktdev *dev, struct pktdev_info *dev_info)
+{
+    struct pmd_lport *lport = dev->data->dev_private;
+
+    dev_info->driver_name = PMD_NET_TAP_NAME;
+    dev_info->if_index    = if_nametoindex(lport->if_name);
```
[the remark this answers is not in the document]
Comment (KeithWiles) [reply lib/core/pmds/net/tap/pmd_tap.c:192]: Need to update af_packet too.

File: lib/core/pmds/net/tap/pmd_tap.c [hunk 4]
```diff
@@ -0,0 +1,381 @@
...
+    int fd;
+    struct iovec *rd;
+
+    uint64_t n_pkts;
+    uint64_t n_bytes;
+};
+
+struct pmd_lport {
+    uint16_t lport_id;
+    char if_name[IFNAMSIZ + 1];
```
[the remark this answers is not in the document]
Comment (KeithWiles) [reply lib/core/pmds/net/tap/pmd_tap.c:48]: Taken from af_packet and that PMD needs to be fixed.

File: lib/core/pmds/net/tap/pmd_tap.c [hunk 5]
```diff
@@ -0,0 +1,381 @@
...
+    pktmbuf_info_t *pi;
+    struct tap_info *ti;
+    struct ether_addr eth_addr;
+
+    struct tap_rx_q *rxq;
+    struct tap_tx_q *txq;
+};
+
+static uint16_t
+pmd_tuntap_rx(void *queue, pktmbuf_t **bufs, uint16_t nb_pkts, int tap_type __cne_unused)
```
[the remark this answers is not in the document]
Comment (KeithWiles) [reply lib/core/pmds/net/tap/pmd_tap.c:58]: I would like to add a few features from I think would be nice from the DPDK PMD, which needed to know the type, plus I do like the consistence. I will remove them for now.

File: lib/core/pmds/net/tap/pmd_tap.c (hunk 1 above)
Comment (KeithWiles) [reply lib/core/pmds/net/tap/pmd_tap.c:146]: Yes, should mask the bits for version.

File: lib/core/pmds/net/tap/pmd_tap.c [hunk 6]
```diff
@@ -0,0 +1,381 @@
...
+    dev_info->driver_name = PMD_NET_TAP_NAME;
+    dev_info->if_index    = if_nametoindex(lport->if_name);
+
+    return 0;
+}
+
+static int
+pmd_tun_dev_info(struct cne_pktdev *dev, struct pktdev_info *dev_info)
+{
+    struct pmd_lport *lport = dev->data->dev_private;
```
[the remark this answers is not in the document]
Comment (KeithWiles) [reply lib/core/pmds/net/tap/pmd_tap.c:200]: Again from af_packet.

File: lib/core/pmds/net/tap/pmd_tap.c [hunk 7]
```diff
@@ -0,0 +1,381 @@
...
+        bufs++;
+
+        pktmbuf_port(m)     = rxq->lport_id;
+        pktmbuf_data_len(m) = len;
+
+        n_rx_pkts++;
+        n_rx_bytes += len;
+    }
+
+    if (n_rx_pkts < nb_cnt)
```
[the remark this answers is not in the document]
Comment (jeffreybshaw) [reply lib/core/pmds/net/tap/pmd_tap.c:102]: I would consider this unusable. Allocating and freeing millions of buffers unnecessarily is really not a good design.

File: lib/core/pmds/net/tap/pmd_tap.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (jeffreybshaw) [reply lib/core/pmds/net/tap/pmd_tap.c]: Let's fix that in a separate PR.

File: lib/core/pmds/net/tap/pmd_tap.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (KeithWiles) [reply lib/core/pmds/net/tap/pmd_tap.c:102]: I will look at this performance improvement.

[the remark this answers is not in the document]
Comment (KeithWiles) [reply lib/core/pmds/net/tap/pmd_tap.c]: I will push an update for this issue.

File: lib/core/pmds/net/tap/pmd_tap.c [hunk 8]
```diff
@@ -0,0 +1,414 @@
...
+    int n_rx_bytes       = 0;
+    struct tun_pi pi;
+    uint16_t len;
+
+    if (!rxq || !bufs || nb_pkts == 0)
+        return 0;
+
+    /* Read packets from the TAP socket & store in allocated mbuf */
+    for (int i = 0; i < nb_pkts; i++) {
+        pktmbuf_t *m = pmd_alloc_rx_mbuf(rxq);
```
Comment (jeffreybshaw) [inline lib/core/pmds/net/tap/pmd_tap.c:97]: I like the scheme you've implemented here.

**Review by jeffreybshaw — APPROVED** [review]

**Pull request merged**